logsmanagement/db_api.c

```diff
@@ -0,0 +1,1398 @@
+        freez((void *) p_file_info->db_dir);\
+        p_file_info->db_dir = strdupz("");\
+        uv_mutex_unlock(p_file_info->db_mut);\
+        uv_rwlock_rdunlock(&p_file_info->circ_buff->buff_realloc_rwlock);\
+        return (void) uv_thread_create(p_file_info->db_writer_thread, db_writer_db_mode_none, p_file_info);\
+    }
+        
+    struct timespec ts_db_write_start, ts_db_write_end, ts_db_rotate_end;
+    while(__atomic_load_n(&p_file_info->state, __ATOMIC_RELAXED) == LOG_SRC_READY){
+        clock_gettime(CLOCK_THREAD_CPUTIME_ID, &ts_db_write_start);
+
+        uv_rwlock_rdlock(&p_file_info->circ_buff->buff_realloc_rwlock);
+        uv_mutex_lock(p_file_info->db_mut);
+        
+        /* ---------------------------------------------------------------------
+         * Read items from circular buffer and store them in disk BLOBs. 
+         * After that, SQLite metadata is updated.
+         * ------------------------------------------------------------------ */
+        Circ_buff_item_t *item = circ_buff_read_item(p_file_info->circ_buff);
+        while (item) {
+            m_assert(TEST_MS_TIMESTAMP_VALID(item->timestamp), "item->timestamp == 0"); 
+            m_assert(item->text_compressed_size != 0, "item->text_compressed_size == 0");
+            m_assert(item->text_size != 0, "item->text_size == 0");
+
+            /* Write logs in BLOB */
+            uv_fs_t write_req;
+            uv_buf_t uv_buf = uv_buf_init((char *) item->text_compressed, (unsigned int) item->text_compressed_size);
+            rc = uv_fs_write(   NULL, &write_req, 
+                                p_file_info->blob_handles[p_file_info->blob_write_handle_offset], 
+                                &uv_buf, 1, blob_filesize, NULL); // Write synchronously at the end of the BLOB file
+            uv_fs_req_cleanup(&write_req);
+            if(unlikely(rc < 0)){
+                throw_error(p_file_info->chart_name, ERR_TYPE_LIBUV, rc, __LINE__, __FILE__, __FUNCTION__);
+                circ_buff_read_done(p_file_info->circ_buff);
+                return_db_writer_db_mode_none(p_file_info);
+            }
+            
+            /* Ensure data is flushed to BLOB via fdatasync() */
+            uv_fs_t dsync_req;
+            rc = uv_fs_fdatasync(   NULL, &dsync_req, 
+                                    p_file_info->blob_handles[p_file_info->blob_write_handle_offset], NULL);
+            uv_fs_req_cleanup(&dsync_req);
+            if (unlikely(rc)){
+                throw_error(p_file_info->chart_name, ERR_TYPE_LIBUV, rc, __LINE__, __FILE__, __FUNCTION__);
+                circ_buff_read_done(p_file_info->circ_buff);
+                return_db_writer_db_mode_none(p_file_info);
+            }
+            
+            if(unlikely(
+                /* Write metadata of logs in LOGS_TABLE */
+                SQLITE_OK != (rc = sqlite3_exec(p_file_info->db, "BEGIN TRANSACTION;", NULL, NULL, NULL)) ||
```

Comment:
Considering that you are having a transaction here, in any possible fail, nothing will be executed after this. Either you are missing a `rollback` or a call for `sqlite3_finalize`.

Reply:
I am resolving after to see the necessary command below.